### PR TITLE
[Distributed Excluded Zone] Handle when depart is in an excluded zone

### PR DESCRIPTION
**Pull request by xlqian**
https://navitia.atlassian.net/browse/NAV-2802

### CODE REVIEW

File: source/jormungandr/jormungandr/excluded_zones_manager.py [hunk 1]
```diff
@@ -67,28 +84,44 @@ def get_excluded_zones(instance_name=None, mode=None):
...
+    )
+    def get_excluded_zones(instance_name=None, mode=None, date=None):
+        excluded_zones = []
+        for json_content in ExcludedZonesManager.get_all_excluded_zones():
+            if instance_name is not None and json_content.get('instance') != instance_name:
+                continue
+            if mode is not None and mode not in json_content.get("modes", []):
+                continue
+            if date is not None and not ExcludedZonesManager.is_activated(
+                json_content['activation_periods'], date
```
Comment (kadhikari) [inline source/jormungandr/jormungandr/excluded_zones_manager.py]: activation_periods is mandatory in json_content ??

File: source/jormungandr/jormungandr/excluded_zones_manager.py [hunk 2]
```diff
@@ -67,28 +84,44 @@ def get_excluded_zones(instance_name=None, mode=None):
...
+            ):
+                continue
+            excluded_zones.append(json_content)
+
+        return excluded_zones
+
+    @classmethod
+    @cache.memoize(app.config[str('CACHE_CONFIGURATION')].get(str('ASGARD_S3_DATA_TIMEOUT'), 1 * 60))
+    def is_excluded(cls, obj, mode, timestamp):
+        print(mode)
```
Comment (kadhikari) [inline source/jormungandr/jormungandr/excluded_zones_manager.py]: ??

Comment (azime) [inline source/jormungandr/jormungandr/excluded_zones_manager.py]:
```suggestion
        print(mode)
```
to remove

File: source/jormungandr/jormungandr/excluded_zones_manager.py [hunk 3]
```diff
@@ -28,15 +28,23 @@
...
 from jormungandr import app, memory_cache, cache
 from jormungandr.resource_s3_object import ResourceS3Object
 
 
 class ExcludedZonesManager:
+    excluded_shapes = dict()  # type: Dict[str, shapely.geometry]
+
     @staticmethod
     @cache.memoize(app.config[str('CACHE_CONFIGURATION')].get(str('ASGARD_S3_DATA_TIMEOUT'), 24 * 60))
     def get_object(resource_s3_object):
```
Comment (azime) [inline source/jormungandr/jormungandr/excluded_zones_manager.py:50]: can you add **_instance_publication_date_**

File: source/jormungandr/jormungandr/excluded_zones_manager.py [hunk 4]
```diff
@@ -67,28 +84,44 @@ def get_excluded_zones(instance_name=None, mode=None):
...
-
         except Exception:
             logger.exception(
-                "Error on fetching excluded zones: bucket: {}, instance: {}, mode ={}",
+                "Error on fetching excluded zones: bucket: {}",
                 bucket_name,
-                instance_name,
-                mode,
             )
+        cls.excluded_shapes = [shapely.wkt.loads(zone.get('shape', '')) for zone in excluded_zones]
```
Comment (azime) [inline source/jormungandr/jormungandr/excluded_zones_manager.py]: **_load_** may not work, no ?

File: source/jormungandr/jormungandr/excluded_zones_manager.py [hunk 5]
```diff
@@ -67,28 +84,44 @@ def get_excluded_zones(instance_name=None, mode=None):
...
 
         return excluded_zones
+
+    @staticmethod
+    @memory_cache.memoize(
+        app.config[str('MEMORY_CACHE_CONFIGURATION')].get(str('ASGARD_S3_DATA_TIMEOUT'), 5 * 60)
+    )
+    def get_excluded_zones(instance_name=None, mode=None, date=None):
```
Comment (azime) [inline source/jormungandr/jormungandr/excluded_zones_manager.py:120]: can you add **_instance_publication_date_**

File: source/jormungandr/jormungandr/excluded_zones_manager.py (hunk 1 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/excluded_zones_manager.py]: yes

File: source/jormungandr/jormungandr/excluded_zones_manager.py (hunk 3 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/excluded_zones_manager.py:50]: for excluded_zones, they have nothing to do with the data of instance, it does not make sense to relate excluded_zones to instance_publication_date

File: source/jormungandr/jormungandr/street_network/asgard.py [hunk 6]
```diff
@@ -242,7 +242,20 @@ def _get_street_network_routing_matrix(
             req.sn_routing_matrix.mode = FallbackModes.car.name
 
         res = self._call_asgard(req, request_id)
+
+        # to handle the case where all origins or all destinations happen to be located in excluded zones
+        # Asgard could have returned a matrix filled with Unreached status, which is kind of waste of the bandwidth
+        # So instead, asgard return with and error_id(all_excluded), we fill the matrix with just on element
```
Comment (xlqian) [inline source/jormungandr/jormungandr/street_network/asgard.py]:
```suggestion
        # So instead, asgard return with an error_id(all_excluded), we fill the matrix with just on element
```

**Review by kadhikari — APPROVED** [review]
Seems fine to me :+1:

**Review by azime — APPROVED** [review]

**Pull request merged**